### PR TITLE
adding dynamic update of LED type dropdown

### PR DESCRIPTION
**Pull request by DedeHai**
removing restriction to only allow changing of last bus by dynamically updating the dropdown menus

@blazoncek is this a valid fix or did I misunderstand the issue?

## Summary by CodeRabbit

* **Improvements**
  * LED type dropdowns now dynamically recalculate and restrict options based on current bus usage and type limits to prevent incompatible selections.
  * Recalculation runs during UI initialization and whenever LED configuration changes, ensuring consistent enforcement across all selectors.
  * Centralized update logic preserves valid selections when adding or changing LEDs for more predictable behavior.

### CODE REVIEW

**Comment by blazoncek** [comment]
Looks like a good solution. 👍 
Except only one call to update is needed at the end of UI() function. And no eventlistener.

File: wled00/data/settings_leds.htm [hunk 1]
```diff
@@ -812,6 +795,32 @@
...
+				if (isVir(t)) virtB++;
+			});
+			// enable/disable type options according to limits in dropdowns
+			LTs.forEach(sel => {
+				const curType = parseInt(sel.value);
+				const disable = (q) => sel.querySelectorAll(q).forEach(o => o.disabled = true);
+				const enable  = (q) => sel.querySelectorAll(q).forEach(o => o.disabled = false);
+				enable('option'); // reset all first
+				// max digital count
+				let maxDB = maxD - ((is32() || isS2() || isS3()) ? (!d.Sf["PR"].checked) * 8 - (!isS3()) : 0);
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/data/settings_leds.htm]: @blazoncek this line is unchanged from previous code. is this the indended behaviour or is the rabbit right?
Chip | PR.checked | Explanation | Effective subtraction | Result
-- | -- | -- | -- | --
ESP32/S2 | false (mono I²S) | Mono mode uses 8 busses | 8 - 1 = 7 | maxDB = maxD - 7
ESP32/S2 | true (parallel) | Parallel mode → don’t reserve I²S busses | 0 - 1 = -1 | maxDB = maxD + 1
ESP32-S3 | false (mono I²S) | Mono mode uses 8 busses | 8 - 0 = 8 | maxDB = maxD - 8
ESP32-S3 | true (parallel) | Only parallel mode available | 0 - 0 = 0 | maxDB = maxD

[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/data/settings_leds.htm]: S2 : 5/12
S3 : 4/12
32 : 9/16
These are supported configurations. First number is with parallel I2S disabled/unchecked.

**Comment by blazoncek** [comment]
Tested this and does not work as expected.
Scenario: Keep adding new buses (i.e. WS281x) and it will not disable unavailable ones.

File: wled00/data/settings_leds.htm (hunk 1 above)
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/data/settings_leds.htm]: The formula above is correct.

**Comment by blazoncek** [comment]
I've come up with the solution:
```c++
			LTs.forEach((sel,i) => {
				const last = i == LTs.length-1;
				const curType = parseInt(sel.value);
				const disable = (q) => sel.querySelectorAll(q).forEach(o => o.disabled = true);
				const enable  = (q) => sel.querySelectorAll(q).forEach(o => o.disabled = false);
				enable('option'); // reset all first
				// max digital count
				let maxDB = maxD - ((is32() || isS2() || isS3()) ? (!d.Sf["PR"].checked) * 8 - (!isS3()) : 0);
				// disallow adding more of a type that has reached its limit
				if (digitalB > maxDB && (!isDig(curType) || last)) disable('option[data-type="D"]');
				if (twopinB > 2 && (!isD2P(curType) || last)) disable('option[data-type="2P"]');
				disable(`option[data-type^="${'A'.repeat(maxA - analogB + (isPWM(curType)?numPins(curType):0) + 1)}"]`);
				// we need to handle last added bus separately to properly select available type
				if (last && sel.options[0].disabled) sel.selectedIndex = sel.querySelector('option:not(:disabled)').index;
			});
```

**Comment by DedeHai** [comment]
> I've come up with the solution

this is much cleaner than the version from my last commit and achieves the same.

**Comment by DedeHai** [comment]
> I've come up with the solution:
> 
> ```c++
> 			LTs.forEach((sel,i) => {
> 				const last = i == LTs.length-1;
> 				const curType = parseInt(sel.value);
> 				const disable = (q) => sel.querySelectorAll(q).forEach(o => o.disabled = true);
> 				const enable  = (q) => sel.querySelectorAll(q).forEach(o => o.disabled = false);
> 				enable('option'); // reset all first
> 				// max digital count
> 				let maxDB = maxD - ((is32() || isS2() || isS3()) ? (!d.Sf["PR"].checked) * 8 - (!isS3()) : 0);
> 				// disallow adding more of a type that has reached its limit
> 				if (digitalB >= maxDB && (!isDig(curType) || last)) disable('option[data-type="D"]');
> 				if (twopinB >= 2 && (!isD2P(curType) || last)) disable('option[data-type="2P"]');
> 				disable(`option[data-type^="${'A'.repeat(maxA - analogB + (isPWM(curType)?numPins(curType):0) + 1)}"]`);
> 				// we need to handle last added bus separately to properly select available type
> 				if (last && sel.options[0].disabled) sel.selectedIndex = sel.querySelector('option:not(:disabled)').index;
> 			});
> ```

this actuall does not work properly, my version with setting to a "virtual bus" does work correctly.

**Comment by blazoncek** [comment]
I initially went with virtual bus too but found it annoying.
I'll poke around.

**Comment by DedeHai** [comment]
your version worked in my tests when removing:
`sel.value = sel.querySelector('option[data-type="N"]').value;`
from my code and in your code remove the last line.

**Comment by DedeHai** [comment]
@netmindz once this is tested a bit more, it can be backported to 0.15

**Comment by blazoncek** [comment]
@DedeHai remove your last commit and move `updateTypeDropdowns()` in `UI()` just after `// enable/disable LED fields` (L:269).

**Comment by blazoncek** [comment]
I've spent a better part of yesterday solving the issue without much succes while trying to keep existing behaviour.

What I did manage to get working in the end is something like @DedeHai did with selecting non-demanding On/Off output when creating a new output.

This output has no restrictions (apart from available GPIO) so can be selected at any time and allows AI generated code to function unchanged. However the place for `updateTypeDropdowns()` is in line 269 as mentioned above.

The unfortunate consequence is that On/Off bus is selected as default when adding a new bus instead of usual 1st available type.

**Comment by netmindz** [comment]
Sadly I can't just swap that base of this PR and if we are going to use this only for 0.15 then we need a fresh pr with these commits cheery picked @DedeHai

**Comment by DedeHai** [comment]
I can create a new PR targeted for 0.15

**Comment by DedeHai** [comment]
@blazoncek is this "good enough" to be merged and backported to 0.15? You did mention that there should be a second place to update the list, aka
``` c++
sel.value = sel.querySelector('option[data-type="N"]').value;
updateTypeDropdowns(); // update valid bus options including this new one
sel.selectedIndex = sel.querySelector('option:not(:disabled)').index;
```

**Comment by blazoncek** [comment]
> is this "good enough"

TBH I never tested your approach though I borrowed some/most from it. Take a peek in my dev branch if you can take something from there. I did notice that selectedIndex = On/Off bus needs to be followed by updateDropdown, then selectedInde = first avaialble and updateDropdown again to work in every circumstance. See my code.

IMO it should be an easy port to 0.15 as I only work on 0.15 codebase.

File: wled00/data/settings_leds.htm [hunk 2]
```diff
@@ -522,18 +513,14 @@
...
+				// temporarily set to virtual (network) type to avoid "same type" exception during dropdown update
 				let sel = d.getElementsByName("LT"+s)[0];
-				// 32 & S2 supports mono I2S as well as parallel so we need to take that into account; S3 only supports parallel
-				let maxDB = maxD - (is32() || isS2() || isS3() ? (!d.Sf["PR"].checked)*8 - (!isS3()) : 0); // adjust max digital buses if parallel I2S is not used
-				if (digitalB >= maxDB) disable(sel,'option[data-type="D"]'); // NOTE: see isDig()
-				if (twopinB  >= 2)     disable(sel,'option[data-type="2P"]'); // NOTE: see isD2P() (we will only allow 2 2pin buses)
-				disable(sel,`option[data-type^="${'A'.repeat(maxA-analogB+1)}"]`); // NOTE: see isPWM()
+				sel.value = sel.querySelector('option[data-type="N"]').value;
+				updateTypeDropdowns(); // update valid bus options including this new one
 				sel.selectedIndex = sel.querySelector('option:not(:disabled)').index;
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/data/settings_leds.htm:520]: this is an edge case that is hard to achieve

**Comment by DedeHai** [comment]
I did test the 0.15 backport and since it is "identical" this PR also gets my green light for merging.
